docs/pages/onBeforeRoute.page.server.mdx

```diff
@@ -1,12 +1,54 @@
+import { HookTypeScriptHints } from '../components/HookTypeScriptHints'
 import { Link, Warning } from '@brillout/docpress'
 
 The `onBeforeRoute()` hook provides full control over routing.
 
-It is usually used for [internationalization](/i18n).
+It is usually used for [internationalization](/i18n):
+
+```js
+export { onBeforeRoute }
+
+async function onBeforeRoute(pageContext) {
+  const { urlWithoutLocale, locale } = extractLocale(pageContext.urlOriginal)
+  return {
+    pageContext: {
+      // We make `locale` available as `pageContext.locale`
+      locale,
+      // We overwrite the original URL
+      urlOriginal: urlWithoutLocale
+    }
+  }
+}
+```
 
 We can also use it to [implement a custom router](https://github.com/vikejs/vike/issues/140).
 
 <Warning>For use cases other than i18n, open a GitHub ticket or reach out on Discord: `onBeforeRoute()` is purposely limited but the limitations can be lifted in order to unlock new capabilities.</Warning>
 
 Example:
 - <Link href='/examples/i18n/renderer/_default.page.route.js' />
+
+## TypeScript
+
+```ts
+export { onBeforeRoute }
+
+import type { OnBeforeRouteAsync } from 'vike/types'
+
+const onBeforeRoute: OnBeforeRouteAsync = async (pageContext): ReturnType<OnBeforeRouteAsync> => {
+  const { urlWithoutLocale, locale } = extractLocale(pageContext.urlOriginal)
+  return {
+    pageContext: {
+      // We make `locale` available as `pageContext.locale`
+      locale,
+      // We overwrite the original URL
+      urlOriginal: urlWithoutLocale
```

Comment:
I'm realizing that TypeScript would compain because `urlOriginal` isn't part of `Vike.PageContext`. So I think we need to add it to the return type of `OnBeforeRoute`.

Reply:
```ts
type PageContextServer = PageContextBuiltInServer & Vike.PageContext

type PageContextBuiltInServer =  { urlOriginal: string, /* ... */ }
```

In other words: `Vike.PageContext` only includes custom stuff defined by the user. Maybe https://vike.dev/pageContext#typescript should be clearer 🤔.

Reply:
> So I think we need to add it to the return type of `OnBeforeRoute`.

done :heavy_check_mark: 

> Maybe https://vike.dev/pageContext#typescript should be clearer 🤔.

I've re-read it, I think it's clear actually, thanks!